### ISSUE TITLE
constant specification confused for enumeration

### ISSUE THREAD

**Issue by cortlandstarrett** [xtuml/pyxtuml#20]
pyxtuml prebuild cannot discriminate a fully qualified constant specification from an enumeration.

Consider a constant specification TimeConversion with constants.

<img width="189" alt="Screen Shot 2020-06-10 at 4 08 22 PM" src="https://user-images.githubusercontent.com/1993198/84313557-aef28580-ab34-11ea-83f5-034e088b5a97.png">

and the following action language.

<img width="489" alt="Screen Shot 2020-06-10 at 4 09 36 PM" src="https://user-images.githubusercontent.com/1993198/84313649-d34e6200-ab34-11ea-847b-bc787b60ff63.png">

The lines commented out cause pyxtuml.prebuild a problem.

Here is the stack trace.  Notice accept_EnumNode.

```
[INFO] Performing pyxtuml pre-build...
[ERROR] Traceback (most recent call last):
[ERROR]   File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 174, in _run_module_as_main
[ERROR]     "__main__", fname, loader, pkg_name)
[ERROR]   File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
[ERROR]     exec code in run_globals
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/bridgepoint/prebuild.py", line 1914, in <module>
[ERROR]     main()
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/bridgepoint/prebuild.py", line 1908, in main
[ERROR]     prebuild_model(m)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/bridgepoint/prebuild.py", line 1873, in prebuild_model
[ERROR]     prebuild_action(inst)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/bridgepoint/prebuild.py", line 1851, in prebuild_action
[ERROR]     return walker.accept(root)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/xtuml/tools.py", line 220, in accept
[ERROR]     r = fn(node, **kwargs)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/bridgepoint/prebuild.py", line 1562, in accept_BodyNode
[ERROR]     return ActionPrebuilder.accept_BodyNode(self, node)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/bridgepoint/prebuild.py", line 439, in accept_BodyNode
[ERROR]     self.accept(node.block.statement_list)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/xtuml/tools.py", line 220, in accept
[ERROR]     r = fn(node, **kwargs)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/bridgepoint/prebuild.py", line 458, in accept_StatementListNode
[ERROR]     act_smt = self.accept(child)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/xtuml/tools.py", line 220, in accept
[ERROR]     r = fn(node, **kwargs)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/bridgepoint/prebuild.py", line 781, in accept_AssignmentNode
[ERROR]     v_val_r = self.accept(node.expression)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/xtuml/tools.py", line 220, in accept
[ERROR]     r = fn(node, **kwargs)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/bridgepoint/prebuild.py", line 1009, in accept_BinaryOperationNode
[ERROR]     v_val_r = self.accept(node.right)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/xtuml/tools.py", line 220, in accept
[ERROR]     r = fn(node, **kwargs)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/bridgepoint/prebuild.py", line 1009, in accept_BinaryOperationNode
[ERROR]     v_val_r = self.accept(node.right)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/xtuml/tools.py", line 220, in accept
[ERROR]     r = fn(node, **kwargs)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/bridgepoint/prebuild.py", line 1109, in accept_EnumNode
[ERROR]     relate(v_val, s_dt, 820)
[ERROR]   File "/Library/Python/2.7/site-packages/pyxtuml-2.1a0-py2.7.egg/bridgepoint/prebuild.py", line 43, in relate
[ERROR]     assert xtuml.relate(*args, **kwargs)
[ERROR] AssertionError
[ERROR] pyxtuml exited with code 1
```

**Comment by john-tornblom** [comment]
I added a new branch at https://github.com/xtuml/pyxtuml/tree/prebuilding-named-constants with a potential fix, but I have no idea if it works since I do not have a test case that triggers the bug.

Could you add one to that branch, preferably one that triggers both new control flows (i.e., including a case where the constant group or constant name is miss-spelled)?

**Comment by cortlandstarrett** [comment]
Sweet.  Very quick response.  Either John Wolfe or I will provide a test
case.  We are working together on the model that revealed the issue.  And,
BTW, we are compiling with Ciera which is depending heavily upon your
work.  :)

**Comment by cortlandstarrett** [comment]
Hi, John.  I branched your branch and added constenum.xtuml to the tests/resources folder.
python -m bridgepoint.prebuild -v resources/constenum.xtuml -o /tmp/constenum.out exhibits the failure.
I looked briefly at how to stitch this in to the automated test but need guidance.

**Comment by cortlandstarrett** [comment]
Prebuild gets further, but there is a still a problem.

File "/Users/cort/git/pyxtuml/bridgepoint/prebuild.py", line 781, in accept_AssignmentNode
    v_val_r = self.accept(node.expression)
  File "xtuml/tools.py", line 220, in accept
    r = fn(node, **kwargs)
  File "/Users/cort/git/pyxtuml/bridgepoint/prebuild.py", line 1122, in accept_EnumOrNamedConstantNode
    relate(v_len, cnst_syc, 850)
UnboundLocalError: local variable 'v_len' referenced before assignment

**Comment by john-tornblom** [comment]
Thanks. Remind me, are constants accessible without the namespace prefix?

**Comment by cortlandstarrett** [comment]
Yes.  And the test case includes that usage.
On the other hand, enumerators must be qualified.

Here is the action language from the test case:

contant_with_qualification = TimeConversion::QuantumPerDay;
contant_without_qualification = QuantumPerDay;
enum_with_qualification = DayOfWeek::Friday;
// enum_without_qualification is not allowed by OAL parser

**Comment by john-tornblom** [comment]
Ah, sorry, just noticed that, thanks!

**Comment by john-tornblom** [comment]
It is a bit tricky to test both qualified and non-qualified constants right now.

There is a round trip test:
  text1 --prebuild--> instances1 --sourcegen--> text2 ---prebuild--> instances2,
followed by:
  assert instances1 == instances2

which does not leverage the structural parts of the model to determine the AST node types.
Consequently, constants appear as variable access and the assertion fails, so I will omit that test for now.

**Comment by john-tornblom** [comment]
I believe there is a working version at https://github.com/xtuml/pyxtuml/tree/prebuilding-named-constants now.

Please confirm, and I'll rebase, squash, and push to master.

**Comment by cortlandstarrett** [comment]
Confirmed.  I can run the test model.  And I can build my original application model now.

**Comment by john-tornblom** [comment]
fixed in 70e2d867
